### PR TITLE
Add warmup for Llama 

### PR DESCRIPTION
**Pull request by digantdesai**
Load the model. Run the everything twice. Reset stats in between two runs. Also decrease logging level for warm up.

Notes:

Tested on Android and Mac. With Llama2 and Llama3 - with temperature=0 produces same output.
This warm up option is disabled by default.
This is inspired from llama.cpp options [[1](https://github.com/ggerganov/llama.cpp/blob/ea9c32be71b91b42ecc538bd902e93cbb5fb36cb/common/common.cpp#L897-L929), [2](https://github.com/ggerganov/llama.cpp/blob/ea9c32be71b91b42ecc538bd902e93cbb5fb36cb/examples/llama-bench/llama-bench.cpp#L1595-L1602)].
Sample [runs](https://www.internalfb.com/phabricator/paste/view/P1613261035)

### CODE REVIEW

**Review by metascroy — APPROVED** [review]

File: examples/models/llama2/runner/runner.h [hunk 1]
```diff
@@ -41,7 +41,11 @@ class Runner {
       std::function<void(const std::string&)> token_callback = {},
       std::function<void(const ::executorch::extension::llm::Stats&)>
           stats_callback = {},
-      bool echo = true);
+      bool echo = true,
+      bool warming = false);
+  ::executorch::runtime::Error warmup(
+      const std::string& prompt,
+      int32_t seq_len = 128);
```
Comment (mcr229) [inline examples/models/llama2/runner/runner.h:48]: should we move seq_len out to a constant variable DEFAULT_SEQ_LEN that is shared by both generate and warmup?

File: examples/models/llama2/runner/runner.cpp [hunk 2]
```diff
@@ -239,26 +254,46 @@ Error Runner::generate(
       prompt_tokens, num_prompt_tokens, seq_len, wrapped_callback));
 
   stats_.inference_end_ms = llm::time_in_ms();
-  printf("\n");
-  ET_LOG(
-      Info,
+  if (!warmup) {
+    printf("\n");
```
Comment (mcr229) [inline examples/models/llama2/runner/runner.cpp:258]: LOG?

File: examples/models/llama2/runner/runner.cpp [hunk 3]
```diff
@@ -161,16 +170,22 @@ Error Runner::generate(
     stats_.model_load_end_ms = llm::time_in_ms();
   }
 
-  ET_LOG(
-      Info,
+  if (warmup) {
+    ET_LOG(Info, "Doing a warmup run...");
+  }
```
Comment (mcr229) [inline examples/models/llama2/runner/runner.cpp:175]: maybe add this before generate in Runner::warmup?

File: .ci/scripts/test_llama.sh [hunk 4]
```diff
@@ -213,7 +213,7 @@ echo "Creating tokenizer.bin"
 $PYTHON_EXECUTABLE -m extension.llm.tokenizer.tokenizer -t tokenizer.model -o tokenizer.bin
 
 
-RUNTIME_ARGS="--model_path=${EXPORTED_MODEL_NAME} --tokenizer_path=tokenizer.bin --prompt=Once --temperature=0 --seq_len=10"
+RUNTIME_ARGS="--model_path=${EXPORTED_MODEL_NAME} --tokenizer_path=tokenizer.bin --prompt=Once --temperature=0 --seq_len=10 --warmup=1"
```
Comment (mcr229) [inline .ci/scripts/test_llama.sh:216]: why warm up in CI?

**Review by mcr229 — APPROVED** [review]
take a look at my nits. otherwise is fine

File: examples/models/llama2/runner/runner.h (hunk 1 above)
Comment (digantdesai) [reply examples/models/llama2/runner/runner.h:48]: so warmup is done on the same prompt/seq_len as a real run today. Default is just something I followed from generate().

File: .ci/scripts/test_llama.sh (hunk 4 above)
Comment (digantdesai) [reply .ci/scripts/test_llama.sh:216]: Yeah good question. I debated this and was thinking about doing two runs and compare outputs w/ and w/o warmup, but CI is expensive so just did w/ warmup and compared output after.

File: examples/models/llama2/runner/runner.cpp [hunk 5]
```diff
@@ -239,26 +254,46 @@ Error Runner::generate(
...
   if (stats_callback) {
     stats_callback(stats_);
   }
 
   return Error::Ok;
 }
 
+Error Runner::warmup(const std::string& prompt, int32_t seq_len) {
+  Error err = generate(
```
Comment (kimishpatel) [inline examples/models/llama2/runner/runner.cpp:286]: It should be prefill, right? not generate? or generate calls prefill? And you are running not warming up by prefill only but entire sequence generation?

also how do you enable this for llava

**Pull request closed without merging**